drivers/clock_control/clock_stm32_ll_u3.c

```diff
@@ -0,0 +1,648 @@
+	case STM32_SRC_HSI16:
+		*rate = STM32_HSI_FREQ;
+		break;
+#endif /* STM32_HSI_ENABLED */
+#if defined(STM32_MSIS_ENABLED)
+	case STM32_SRC_MSIS:
+		*rate = get_msis_frequency();
+		break;
+#endif /* STM32_MSIS_ENABLED */
+#if defined(STM32_MSIK_ENABLED)
+	case STM32_SRC_MSIK:
+		*rate = get_msik_frequency();
+		break;
+#endif /* STM32_MSIK_ENABLED */
+#if defined(STM32_HSE_ENABLED)
+	case STM32_SRC_HSE:
+		*rate = STM32_HSE_FREQ;
+		break;
+#endif /* STM32_HSE_ENABLED */
+#if defined(STM32_LSE_ENABLED)
+	case STM32_SRC_LSE:
+		*rate = STM32_LSE_FREQ;
+		break;
+#endif /* STM32_LSE_ENABLED */
+#if defined(STM32_LSI_ENABLED)
+	case STM32_SRC_LSI:
+		*rate = STM32_LSI_FREQ;
+		break;
+#endif /* STM32_LSI_ENABLED */
+#if defined(STM32_HSI48_ENABLED)
+	case STM32_SRC_HSI48:
+		*rate = STM32_HSI48_FREQ;
+		break;
+#endif /* STM32_HSI48_ENABLED */
+	default:
+		return -ENOTSUP;
+	}
+
+	if (pclken->div) {
+		*rate /= pclken->div + 1;
+	}
+
+	return 0;
+}
+
+static enum clock_control_status stm32_clock_control_get_status(const struct device *dev,
+								clock_control_subsys_t sub_system)
+{
+	struct stm32_pclken *pclken = (struct stm32_pclken *)sub_system;
+
+	ARG_UNUSED(dev);
+
+	if (IN_RANGE(pclken->bus, STM32_PERIPH_BUS_MIN, STM32_PERIPH_BUS_MAX)) {
+		/* Gated clocks */
+		if ((sys_read32(DT_REG_ADDR(DT_NODELABEL(rcc)) + pclken->bus) & pclken->enr)
+		    == pclken->enr) {
+			return CLOCK_CONTROL_STATUS_ON;
+		} else {
+			return CLOCK_CONTROL_STATUS_OFF;
+		}
+	} else {
+		/* Domain clock sources */
+		if (enabled_clock(pclken->bus) == 0) {
+			return CLOCK_CONTROL_STATUS_ON;
+		} else {
+			return CLOCK_CONTROL_STATUS_OFF;
+		}
+	}
+}
+
+static DEVICE_API(clock_control, stm32_clock_control_api) = {
+	.on = stm32_clock_control_on,
+	.off = stm32_clock_control_off,
+	.get_rate = stm32_clock_control_get_subsys_rate,
+	.get_status = stm32_clock_control_get_status,
+	.configure = stm32_clock_control_configure,
+};
+
+static void set_regu_voltage(uint32_t hclk_freq)
+{
+	if (hclk_freq < MHZ(48)) {
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE2);
+	} else {
+		LL_PWR_SetRegulVoltageScaling(LL_PWR_REGU_VOLTAGE_SCALE1);
+	}
+}
+
+static void enable_epod_booster(void)
+{
+	LL_RCC_SetEPODBoosterClkSource(LL_RCC_EPODBOOSTCLKSRCE_MSIS);
+	LL_RCC_SetEPODBoosterClkPrescaler(LL_RCC_EPODBOOSTCLKPRESCAL_1);
+	LL_PWR_EnableEPODBooster();
+	while (LL_PWR_IsActiveFlag_BOOST() == 0) {
+	}
+}
+
+static void configure_clock_with_calibration(int range)
+{
+	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_MSIS);
+
+	/* Wait till System clock is ready */
+	while (LL_RCC_GetSysClkSource() != LL_RCC_SYS_CLKSOURCE_MSIS) {
+	}
+
+	LL_RCC_SetAHBPrescaler(LL_RCC_HCLK_SYSCLK_DIV_1);
+	LL_RCC_SetAPB1Prescaler(LL_RCC_APB1_HCLK_DIV_1);
+	LL_RCC_SetAPB2Prescaler(LL_RCC_APB2_HCLK_DIV_1);
+	LL_RCC_SetAPB3Prescaler(LL_RCC_APB3_HCLK_DIV_1);
+
+	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIx_ENABLED),
+		"MSIx requires LSE clock to be enabled for auto-calibration");
```

Comment:
My suggestion wasn't clear, but you need to replace the `x` with instance name :sweat_smile: (unless I missed something and `STM32_MSIx_ENABLED` is a real macro??)
```suggestion
	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIK_ENABLED),
		"MSIK requires LSE clock to be enabled for auto-calibration");
	BUILD_ASSERT(STM32_LSE_ENABLED || !IS_ENABLED(STM32_MSIS_ENABLED),
		"MSIS requires LSE clock to be enabled for auto-calibration");
```

\+ build assert can be moved at top of file

Reply:
yes, the two macros are defined in stm32_clock_control.h:

#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msik), st_stm32u3_msi_clock, okay)
#define STM32_MSIK_ENABLED	1

#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u5_msi_clock, okay) || \
	DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk_msis), st_stm32u3_msi_clock, okay)
#define STM32_MSIS_ENABLED	1